### ISSUE TITLE
ViewportTexture.get_data() race condition

### ISSUE THREAD

**Issue by msc509** [godotengine/godot#44866]
**Godot version: 3.2.3 mono**
(Mono is not actually used)

**OS/device including version: Pop! OS 20.04 x64**
GPU: GTX 1080 Mobile
Driver: Nvidia 455
Backend: GLES3

**Issue description:**
When trying to read from a `Viewport` by using its associated `ViewportTexture`, sometimes black image is retrieved.

**Steps to reproduce:**
I could not reproduce it with a small example. 
This happens in a substantial project where I have multiple small viewports in the same world, each with one camera.
Viewport textures are displayed in `TexturedRect` of UI, and texture data are read in scripts for process.
Often times black image is returned.
Forcing update by repeatedly setting `Viewport.UPDATE_ONCE` reduces the number of black images, so I suspect this is a race condition.

A refactor later somehow made the issue disappear, and an attempt to reproduce with small project does not have this issue.

**Comment by dalexeev** [comment]
See #44529.

**Comment by Calinou** [comment]
Try waiting for one frame before capturing the image: https://github.com/godotengine/godot-demo-projects/blob/master/viewport/screen_capture/screen_capture.gd

**Comment by msc509** [comment]
> Try waiting for one frame before capturing the image: https://github.com/godotengine/godot-demo-projects/blob/master/viewport/screen_capture/screen_capture.gd

In my original case, I capture on every frame. Some of those succeed, but most fail. Is that what `frame_post_draw` is for?

**Comment by Xrayez** [comment]
Yes, I think in most cases you have to wait until `frame_post_draw` signal is emitted. You may also want to look at the method of drawing without having to rely on `frame_post_draw` signal, as seen in https://gist.github.com/clayjohn/80d9b1e52e1b9a23445504e85a8d7890, but not sure if this would be practical for your use case.
